modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsRequestCollection.java

```diff
@@ -0,0 +1,343 @@
+import org.apache.ignite.internal.processors.query.stat.messages.StatsCollectionResponse;
+import org.apache.ignite.internal.processors.query.stat.messages.StatsKeyMessage;
+import org.apache.ignite.internal.processors.query.stat.messages.StatsObjectData;
+import org.apache.ignite.internal.util.typedef.F;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Statistics collection routine.
+ */
+public class IgniteStatisticsRequestCollection {
+    /** Current collections, collection id to collection status map. */
+    private final Map<UUID, StatCollectionStatus> currColls = new ConcurrentHashMap<>();
+
+    /** Schema manager. */
+    private final SchemaManager schemaMgr;
+
+    /**
+     * Constructor.
+     *
+     * @param schemaMgr Schema manager.
+     */
+    public IgniteStatisticsRequestCollection(SchemaManager schemaMgr) {
+        this.schemaMgr = schemaMgr;
+    }
+
+    /**
+     * Add new statistics collection status.
+     *
+     * @param status status to add.
+     */
+    public void addCollection(StatCollectionStatus status) {
+        currColls.put(status.colId(), status);
+    }
+
+    /**
+     * Thread safe update of statistics collection.
+     *
+     * @param colId statistics collection.
+     * @param transformation transformation, if return {@code null} - status will be removed.
+     */
+    public void updateCollection(UUID colId, Function<StatCollectionStatus, StatCollectionStatus> transformation) {
+        currColls.compute(colId, (k, v) -> transformation.apply(v));
+    }
+
+    /**
+     * Get collection status.
+     *
+     * @param colId Collection id.
+     * @return Collection status.
+     */
+    public StatCollectionStatus getCollection(UUID colId) {
+        return currColls.get(colId);
+    }
+
+
+    /**
+     * Extract groups of stats keys.
+     *
+     * @param keys Statistics key to extract groups from.
+     * @return Map of <group ids></group> to <collection of keys in groups>
+     * @throws IgniteCheckedException In case of lack some of specified objects.
+     */
+    protected Map<CacheGroupContext, Collection<StatsKeyMessage>> extractGroups(Collection<StatsKeyMessage> keys)
+            throws IgniteCheckedException {
+        Map<CacheGroupContext, Collection<StatsKeyMessage>> res = new HashMap<>(keys.size());
+        for (StatsKeyMessage key : keys) {
+            GridH2Table tbl = schemaMgr.dataTable(key.schema(), key.obj());
+            if (tbl == null)
+                throw new IgniteCheckedException(String.format("Can't find object %s.%s", key.schema(), key.obj()));
```

Comment:
What if one of cache was stopped, but other are operational?
Will statistics task result be thrown away or task crashed?

Reply:
For now - whole task should be cancelled. I'll check it...